### ISSUE TITLE
Class Diagram (First Version)

### ISSUE THREAD

**Issue by akaogluozgur** [bounswe/bounswe2018group9#85]
**This is the first draft. Some methods and properties will be added. Especially CRUD operations getter and setters, easy properties such as name etc.**
@byklyci Firstly what comes to my mind are:

---

### Properties:
* None
### Methods:
* getTrending() (All Users are able to see Trending Req: **1.1.1.6**)
* search() (All users can search **1.1.1.4**)
* getUserEvents() (All Users are able to see member's evets. Req: **1.1.1.1**)

---

### Properties:
* Name (Req: **1.1.2.1**)
* Surname (Req: **1.1.2.1**)
* Age (Req: **1.1.2.1**)
* Nickname (Req: **1.1.2.1**)
* E-mail (Req: **1.1.2.1**)
* Password (Req: **1.1.2.1**)
* Profile(Object) (Req: **1.1.3.1.1.**)
* Settings(Object)
* Followers ( signedUser array. Requirements **1.2.8.3.1**)
* Feed (Object: Every signedUser will have a feed based on type of the user.)
* lastPasswordChange (Date. Requirement **2.6.1**)
### Is able to:
* see events created by other signed users
* post comments
* can vote
* see feed page
### Methods:
changePassword() (Requirements **1.1.2.2.3**)
getFollowers(signedUser array. Requirements **1.2.8.3.1**)
follow()  (Requirements **1.2.8.3.1**)
confirmAccount() (Requirement: **1.2.10.1**)
getTimeline()

---

## Admin(Class inherits SignedUser):
__We need to specify what exactly this guy does.UNTIL FINAL DRAFT__
### Properties
* None
### is able to
* manage the system
* delete event

---

### Property:
None.
### is able to:
*  enter their username and password to login to the system.(Req:**1.1.2.2.1.**)
*  have the option to Sign in via their Facebook account.(Req:**1.1.2.2.2.**)
*  have an option to reset their password.(Req:**1.1.2.2.3.**)
*  edit their activity preferences.(Req:**1.1.3.1.2.**)
* search with Semantic, Content and Location options.(Req:**1.1.7.1.**)
### Method:
changePassword() (Requirements 1.1.2.2.3)
getFollowers(signedUser array. Requirements 1.2.8.3.1)
follow() (Requirements 1.2.8.3.1)
confirmAccount() (Requirement: 1.2.10.1)
getTimeline()
createPage()(Requirenment 1.1.6.1.1.)

---

### Property:
We need to decide 
### is able to:
*  see events created by the members(Req: **1.1.1.1.**)
* to search for public content(Req: **1.1.1.4.**)
*see the Trending page when they open the system.(Req: **1.1.1.6.**)
### Method:
* getTrending() (All Users are able to see Trending Req: **1.1.1.6**)
* search() (All users can search **1.1.1.4**)
* getUserEvents() (All Users are able to see member's evets. Req: **1.1.1.1**)
---

### Property:
* location
* price
* name
* willAttendUser (Req: **1.2.6.1**, **1.2.6.3**) 
* maybeAttendUser  (Req: **1.2.6.1**, **1.2.6.3**) 
* notAttendUser  (Req: **1.2.6.1**, **1.2.6.3**) 
* attendedUser (Req: **1.2.6.2**)
* notAttendedUser (Req: **1.2.6.2**)
* commentList
* voteList
### Method:
create()
delete()
edit()

----

### Property:
notification_type (Requirements **1.2.9.2**)
from(signedUser object Requirements **1.2.9.2.2** **1.2.9.2.3** **1.2.9.2.4**)
event(event object Requirements **1.2.9.2.1**) 
### Method:
notify()  (Requirements **1.2.9.2**)

----

### Property:
voteList
user
### Methods:
delete()
edit()

---

### Properties:
isPrivate (boolean for private profiles. Requirement: **1.2.10.3**, **2.7.1.1** )
theme(Requirement: **2.1.4** )
language (Requirement: **2.12.1**)
### Methods:
changeTheme() (Requirement **2.1.4** )
changePrivate() (Requirement: **1.2.10.3** )
changeLanguage() (Requirement: **2.12.1**)

---

### Properties:
* name

---

### Property:
eventList (array of events to be displayed:Requirements **1.2.11.1.2** and **1.2.11.1.3**)
feedType(Requirements **1.2.11.1.1.**)
### Method:
update() (updates the content of the eventList Requirements :**1.2.11.2** and **1.2.11.3**)

---

### Property:
blockedUsers (this signedUsers will be prevented from seeing this vote.Requirement: **1.2.8.3.1**)
votes (vote list. Requirements: **1.2.8.5.1**)
### Method
hide() (Requirements: **1.2.8.5.1**)

---

### Properties:
* searchType (Req: **1.2.1.1**)
### Methods:
* search() (Req: **1.2.1**,  **1.2.1.2**)
---

### Properties:
* notification (req *1.2.2*)
### Methods:
* recommend(signedUser) ( Req **1.2.2**)

---

### Properties
parameterList (Req: **1.2.3.3**)
### Methods
rank(User) (Req: **1.2.3.3**)

---
## Unmet Requirements
* **1.2.10.1** (this does not imply a property, method or an object)
* **1.2.10.2** (this is not a real property of settings it will be handled on front-end)
* **1.2.9.1 ** (this does not imply a property, method or an object)
* **1.2.9.3** (this does not imply a property, method or an object)
* **1.2.8.1** (since we decide we do not include objects from data access layer)
* **1.2.8.2** (since we decide we do not include objects from data access layer)
* **1.2.8.4.1** (what does this requirement even mean?)
* **1.2.7.1** (since we decide we do not include objects from data access layer)
* **1.2.7.2** (since we decide we do not include objects from data access layer)
* **1.2.7.3** (since we decide we do not include objects from data access layer)
* **1.2.2** (system objects are not determined.)
* **1.2.3** (system objects are not determined.)
* **1.2.4** (admin panel is not specified enough)
* **1.2.5** (will be added to various objects)
* non-functional requirements other than **2.1.4** , **2.6.1**, **2.7.1.1** and **2.12.1**
---

**Comment by akaogluozgur** [comment]
@byklyci feel free to object any of them, I will add as I scan the whole requirements. After we finalize this we can draw it.

**Comment by byklyci** [comment]
I am also reading the requirements again. :+1:

**Comment by byklyci** [comment]
**Event(class)**
Type
Name 
Location
Attendees
Tags
Description
Date ,Time
Comment 
Price
Concert Report 
Ranking
Also there will be subclasses about the event.For example the location,comment,tag,rating etc will became a sub of the class and we will write sth under them.

**Comment by byklyci** [comment]
**Comment(class)**
Comment id 
Owner of the comment
Event name tag
Content

**Comment by byklyci** [comment]
I changed Comment Class; Comment(Class inherits Event)

**Comment by akaogluozgur** [comment]
@byklyci I think Comment class should not inherit event but event class has property as commentList

**Comment by byklyci** [comment]
@akaogluozgur your method is so good, i will add also emphasize the requirenments.

**Comment by byklyci** [comment]
I am agree with you about the comment class issue.My fault.

**Comment by akaogluozgur** [comment]
We can change the hierarchy like this. Admin inherits Member, Member inherits Guest since admin can do all of the stuff what member will do, a member can do all of the stuff guests can do

**Comment by byklyci** [comment]
## Guest(Class inherits User):
### Property:
We need to decide 
### is able to:
*  see events created by the members(Req: **1.1.1.1.**)
* to search for public content(Req: **1.1.1.4.**)
*see the Trending page when they open the system.(Req: **1.1.1.6.**)
### Method:
* search() (All users can search **1.1.1.4**)

**Comment by byklyci** [comment]
It can be possible but admin is an admin.In my opinion admin can be a seperated class but i also agree with you, admin can do all of the staff what members will do.I think we can discuss this topic until the final draft. @akaogluozgur

**Comment by akaogluozgur** [comment]
Most of the things are seems done. Let's divide the requirements. I checked the almost all of the non-functional requirements so there is no need to divide them. Let's you check User Requirements and I check  System requirements.

**Comment by byklyci** [comment]
I will handle User Requirenments. :+1:

**Comment by akaogluozgur** [comment]
I will do these after you edit not to override you:
search() req 1.2.1.2
add recommender object
add parameters notification (req 1.2.2)
add methods recommend(signedUser) ( Req 1.2.2)
add ranker object
add propery parameter req 1.2.3.3
add method rank(User)
add 1.2.4req unmet reqs admin panel needs to be designed
add 1.2.5req unmet reqs discuss about annotations

**Comment by okyksl** [comment]
@akaogluozgur @byklyci can we start to integrate these into the `Wiki` under the `Architecture Design`?

**Comment by akaogluozgur** [comment]
@okyksl after we finalize it, it can be done. Tonight and tomorrow I will check for all requirements and try to add forgotten methods and properties.

**Comment by kemaltulum** [comment]
@akaogluozgur @byklyci I have some feedback for you! While designing sequence diagrams, we need functional units that achieve basic jobs. For example, when the user wants to create an event, there should some entity dealing with this request. Right now, I see the `Event` entity with create method, which is nice but, I think, this method should take the details of the event as parameters, or as a whole object like `EventDetails`.

In short, each entity should have such logical methods so that we can utilize them.

**Comment by akaogluozgur** [comment]
@kemaltulum As I remember, In the meeting we did decide that we should not include methods from the presentation and data access layers like in the spring layered architecture, we should only decide general object plan of our project. What I understand is your suggestion is a method from the first layer which redirects the incoming requests to the business layer. If we include such layers we need to change various objects. **However, if you mean only CRUD methods of the individual objects, they will be updated with proper parameters until night.**

**Comment by byklyci** [comment]
@okyksl after the final version of class diagram we can add all into  the Wiki` under the `Architecture Design`

**Comment by kemaltulum** [comment]
Although we do not have a layered structure like Spring framework, we should have some entities like `EventManager` `UserManager`, `RecommendationManager`. What these entities should deal with is another issue but basically they perform the operations needed in our system.

For example, `EventManager` can deal with creation, validation and update of `Event`s.

**Comment by serkan-ozel-2015400123** [comment]
@akaogluozgur  "can edit their activity preferences.(Req:1.1.3.1.2.)" this is from member class
Can you add this to functions of the class as well?

**Comment by nazmicancalik** [comment]
@byklyci @akaogluozgur Could you guys add me to the platform you are creating the class diagrams so that I can help :D

**Comment by byklyci** [comment]
Please talk in the new issue #92. Lets finalize this job guys.

**Comment by okyksl** [comment]
@byklyci will we have a separate [Classes](https://github.com/bounswe/bounswe2018group9/wiki/Classes) page or only a [Class Diagrams](https://github.com/bounswe/bounswe2018group9/wiki/Class-Diagrams) page?

**Comment by byklyci** [comment]
@okyksl We will add tomorrow, both of them(Classes and Class Diagrams page).

**Comment by okyksl** [comment]
I believe the discussion will be continue from #92 as the first version of the Class Diagrams are completed.
